python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,810 @@
+
+            messages = context.allGather(message=str(_rabit_args))
+            _rabit_args = _get_args_from_message_list(messages)
+            evals_result = {}
+            with RabitContext(_rabit_args, context):
+                booster = worker_train(params=booster_params,
+                                       dtrain=dtrain,
+                                       evals=dval,
+                                       evals_result=evals_result,
+                                       **kwargs_params)
+            context.barrier()
+
+            if use_external_storage:
+                shutil.rmtree(external_storage_path_prefix)
+            if context.partitionId() == 0:
+                yield pd.DataFrame(
+                    data={'booster_bytes': [cloudpickle.dumps(booster)]})
+
+        result_ser_booster = dataset.mapInPandas(
+            _train_booster,
+            schema='booster_bytes binary').rdd.barrier().mapPartitions(
+            lambda x: x).collect()[0][0]
+        result_xgb_model = self._convert_to_model(
+            cloudpickle.loads(result_ser_booster))
+        return self._copyValues(self._create_pyspark_model(result_xgb_model))
+
+    def _fit(self, dataset):
+        self._validate_params()
+        # Unwrap the VectorUDT type column "feature" to 4 primitive columns:
+        # ['features.type', 'features.size', 'features.indices', 'features.values']
+        features_col = col(self.getOrDefault(self.featuresCol))
+        label_col = col(self.getOrDefault(self.labelCol)).alias('label')
+        features_array_col = vector_to_array(features_col, dtype="float32").alias("values")
+        select_cols = [features_array_col, label_col]
+
+        has_weight = False
+        has_validation = False
+        has_base_margin = False
+
+        if self.isDefined(self.weightCol) and self.getOrDefault(
+                self.weightCol):
+            has_weight = True
+            select_cols.append(
+                col(self.getOrDefault(self.weightCol)).alias('weight'))
+
+        if self.isDefined(self.validationIndicatorCol) and \
+                self.getOrDefault(self.validationIndicatorCol):
+            has_validation = True
+            select_cols.append(
+                col(self.getOrDefault(
+                    self.validationIndicatorCol)).alias('validationIndicator'))
+
+        if self.isDefined(self.baseMarginCol) and self.getOrDefault(
+                self.baseMarginCol):
+            has_base_margin = True
+            select_cols.append(
+                col(self.getOrDefault(self.baseMarginCol)).alias("baseMargin"))
+
+        dataset = dataset.select(*select_cols)
+        # create local var `xgb_model_creator` to avoid pickle `self` object to remote worker
+        xgb_model_creator = self._get_xgb_model_creator()  # pylint: disable=E1111
+        fit_params = self._gen_fit_params_dict()
+
+        if self.getOrDefault(self.num_workers) > 1:
```

Comment:
May I ask why needs to consider different situations for num_workers? Could we just use `_fit_distributed` ?

Reply:
If num_workers == 1, we don't need to run rabit workers, and we can directly run xgboost single node routine. It would be faster. Is my understanding correct ?

Reply:
It should make no difference for the internal of XGBoost.

Reply:
Looks like the RABIT tracker is running on the the same process as Python driver, If there is only one RABIT worker, IMO, it won't cause the performance issue. Instead, we need to maintain two "same" branches.